### PR TITLE
Attempt to fix the webauthn migration again - part 3 (#18770)

### PR DESCRIPTION
**Pull request by zeripath**
Backport #18770 

v208.go is seriously broken as it misses an ID() check. We need to no-op and remigrate all of the u2f keys.

See #18756

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

File: models/migrations/v209.go [hunk 1]
```diff
@@ -5,140 +5,13 @@
...
-		}
-
-		if len(regs) < 50 {
-			break
-		}
-		start += 50
-		regs = regs[:0]
-	}
+	// no-op
+	// MariaDB asserts that 408 characters is too long for a VARCHAR(410) so this migration is clearly wrong.
```
Comment (zeripath) [inline models/migrations/v209.go]:
```suggestion
	// V208 is badly broken
```

File: models/migrations/v210.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+				UpdatedUnix:     reg.UpdatedUnix,
+				CreatedUnix:     reg.CreatedUnix,
+			}
+
+			has, err := x.ID(reg.ID).Where("id = ?", reg.ID).Get(new(webauthnCredential))
+			if err != nil {
+				return fmt.Errorf("unable to get webauthn_credential[%d]. Error: %v", reg.ID, err)
+			}
+			if !has {
+				_, err = x.ID(remigrated.ID).AllCols().Insert(remigrated)
```
Comment (zeripath) [inline models/migrations/v210.go]:
```suggestion
				_, err = x.Insert(remigrated)
```

**Review by fnetX — APPROVED** [review]
I tested this backport on our data, and I can at least confirm that the resulting data (id, credential_id) in webauthn_credential is equal to what the previous migration was supposed to do :)

I'll probably check later if my key is accepted, but I'll first need to learn how the access the web interface via multiple agent forwarded ssh connections ...

File: models/migrations/migrations.go [hunk 3]
```diff
@@ -369,9 +369,11 @@ var migrations = []Migration{
 	// v207 -> v208
 	NewMigration("Add webauthn table and migrate u2f data to webauthn", addWebAuthnCred),
```
Comment (fnetX) [inline models/migrations/migrations.go]: Does the migration compare against keys that were removed in the meantime? If not, wouldn't it be now a good idea to let 207 only create the table without writing actual data to speed up things a bit?

Comment (zeripath) [reply models/migrations/migrations.go]: OK I've no-oped that too.

**Review by Gusted — APPROVED** [review]

**Review by jolheiser — APPROVED** [review]

**Pull request merged**